### PR TITLE
Code Health - Reworked bio_enc tests to use new test framework.

### PR DESCRIPTION
**Pull request by jon-oracle**
Converted the bio_enc tests to use new test framework. Also reworked tests to do both encrypt and decrypt, and added a few more ciphers.

- [x] tests are added or updated

### CODE REVIEW

File: test/bio_enc_test.c [hunk 1]
```diff
@@ -12,127 +12,199 @@
...
+        BIO_push(b, BIO_new_mem_buf(inp, DATA_SIZE));
         memset(out, 0, sizeof(out));
         out[i] = ~ref[i];
         len = BIO_read(b, out, i);
         /* check for overstep */
-        if (out[i] != (unsigned char)~ref[i]) {
-            fprintf(stderr, "CBC output overstep@%d\n", i);
-            return 1;
-        }
+        if (!TEST_uchar_eq(out[i], (unsigned char)~ref[i]))
```
Comment (mattcaswell) [inline test/bio_enc_test.c]: Probably a "TEST_info()" call might be appropriate to tell us the value of i in the event of a failure.

File: test/bio_enc_test.c [hunk 2]
```diff
@@ -12,127 +12,199 @@
...
+            return 0;
+
         len += BIO_read(b, out + len, sizeof(out) - len);
         BIO_free_all(b);
 
-        if (len != lref || memcmp(out, ref, len)) {
-            fprintf(stderr, "CBC output mismatch@%d\n", i);
-            return 2;
-        }
+        if (!TEST_mem_eq(out, len, ref, lref))
```
Comment (mattcaswell) [inline test/bio_enc_test.c]: Similarly it would be useful to know the value of i here.

File: test/bio_enc_test.c [hunk 3]
```diff
@@ -12,127 +12,199 @@
...
-     * Exercise CBC cipher
-     */
+    /* Encrypt tests */
 
     /* reference output for single-chunk operation */
     b = BIO_new(BIO_f_cipher());
-    if (!BIO_set_cipher(b, EVP_aes_128_cbc(), key, NULL, 0))
-        return -1;
-    BIO_push(b, BIO_new_mem_buf(inp, sizeof(inp)));
+    if (!TEST_int_eq(BIO_set_cipher(b, cipher, key, iv, ENCRYPT), 1))
```
Comment (jon-oracle) [inline test/bio_enc_test.c]: Yes definitely. Not sure what I was thinking...

File: test/bio_enc_test.c [hunk 4]
```diff
@@ -12,105 +12,137 @@
...
-        if (len != lref || memcmp(out, ref, len)) {
-            fprintf(stderr, "CBC output mismatch@%d\n", i);
-            return 2;
+        if (!TEST_mem_eq(out, len, ref, lref)) {
+            TEST_info("Encrypt compare failed @ operation %d", i);
+            return 0;
         }
     }
-
+
```
Comment (mattcaswell) [inline test/bio_enc_test.c]: Spurious whitespace on this line

**Comment by levitte** [comment]
I'll merge this

**Comment by levitte** [comment]
Merged.  I squashed it all into one commit, dd94c37a5c2f783102b125c620000b9719c662d3

File: test/bio_enc_test.c [hunk 5]
```diff
@@ -64,53 +88,61 @@ int main()
...
     b = BIO_new(BIO_f_cipher());
-    if (!BIO_set_cipher(b, EVP_aes_128_ctr(), key, NULL, 0))
-         return -1;
-    BIO_push(b, BIO_new_mem_buf(inp, sizeof(inp)));
-    lref = BIO_read(b, ref, sizeof(ref));
+    if (!TEST_true(BIO_set_cipher(b, cipher, key, iv, DECRYPT)))
+        return 0;
+    /* Use original reference output as input */
+    BIO_push(b, BIO_new_mem_buf(ref, lref));
+    (void)BIO_flush(b);
```
Comment (vdukhovni) [inline test/bio_enc_test.c:116]: Is there a reason for this `BIO_flush()` it is by now in OpenSSL 3.x definitely not correct, but was there a plausible reason for this back when introduced?

Comment (jon-oracle) [reply test/bio_enc_test.c:116]: No idea, it's been some time since I wrote this. I agree it doesn't need to be there though.

**Pull request closed without merging**